### PR TITLE
feat: Initial support for httpx

### PR DESCRIPTION
**Pull request by lmilbaum**
A followup of #1036
Two clients defined: `requests` and `httpx`. As long as `httpx` is not fully implemented and adopted by the community, keeping the `requests` client as the default.

### CODE REVIEW

File: requirements.txt
Comment (nejch) [inline requirements.txt:1]: Will also need an `extra_requires` entry in setup.py, so people ensure they have it with `pip install python-gitlab[httpx]`.

File: gitlab/clients/httpx_client.py
Comment (nejch) [inline gitlab/clients/httpx_client.py:1]: Not sure if we need separate modules at this point as they're quite small.

File: gitlab/clients/httpx_client.py
Comment (nejch) [inline gitlab/clients/httpx_client.py:6]: I would follow camelcase and make them private here, so people know it's an internal implementation only :)

```suggestion
class _HttpxClient:
```

File: gitlab/client.py
Comment (nejch) [inline gitlab/client.py:10]: This should probably be in a try/except clause and we should only throw an error if people explicitly want to use httpx. We don't want to pull in a new dependency by default if not used, IMO.

I did something similar for graphql here:

https://github.com/python-gitlab/python-gitlab/blob/b8ca5ffb69dce6a7f2f1eb07aeec2216d6817bc5/gitlab/client.py#L1248-L1258

File: gitlab/client.py
Comment (nejch) [inline gitlab/client.py:116]: Maybe just `self._client` is enough :) as they are mostly API-compatible for making requests.

But I think we will need some more overriding based on multipart-encoding in some of our helper methods. I would try to make a matrix in functional test where we run them with httpx and then you should start seeing some issues with that.

**Review by nejch — COMMENTED** [review]
@lmilbaum I made a quick initial pass. Would be interested to see how this runs on functional tests. Thanks a lot for working on this, it's something I've been thinking of myself for a while!

File: gitlab/client.py
Comment (nejch) [inline gitlab/client.py:27]: If we do this, it also needs to be a conditional dependency in setup.py when python =< 3.7.

File: requirements.txt
Comment (lmilbaum) [reply requirements.txt:1]: Added to the "install_requires" key where the reqests package is listed

File: gitlab/clients/httpx_client.py
Comment (lmilbaum) [reply gitlab/clients/httpx_client.py:1]: The intention is moving the functionality to the clients classes.  This PR will lay the ground for the coming changes.

File: gitlab/clients/httpx_client.py
Comment (lmilbaum) [reply gitlab/clients/httpx_client.py:6]: Done

File: gitlab/client.py
Comment (lmilbaum) [reply gitlab/client.py:10]: Makes sense. Just on quick question. How can we know upfront (before the package is installed) if a user wants to use one of the packages and not the other?

File: gitlab/client.py
Comment (lmilbaum) [reply gitlab/client.py:116]: I am not sure I know what `multipart-encoding` is

File: gitlab/client.py
Comment (lmilbaum) [reply gitlab/client.py:27]: OK. Worth adding a functional test which checks that. I've implemented something similar in a different project where I used a container to install the package and use it as the environment under test.

File: gitlab/client.py
Comment (nejch) [reply gitlab/client.py:11]: I think we can import at the top initially with a try/except and then later when it needs to be used (e.g. when the user provides the httpx client), if it is `None` we can raise an ImportError.
```python
try:
    import httpx
except ImportError:
    httpx = None

# ...

if not httpx:
     raise ImportError("... please install with python-gitlab[httpx]")
```

We do something similar in our cli modules:
https://github.com/python-gitlab/python-gitlab/blob/9b88132078ed37417c2a45369b4976c9c67f7882/gitlab/v4/cli.py#L421-L430

httpx itself is actually a good example of this kind of pattern with all its optional transports:

https://github.com/encode/httpx/blob/14a1704be507cf435fa84da75c249c2ecd1c3981/httpx/_transports/default.py#L158-L164

File: gitlab/client.py
Comment (nejch) [reply gitlab/client.py]: Let's see when we start running functional tests with httpx. I think we will run into issues then and we can debug.

File: requirements.txt
Comment (nejch) [reply requirements.txt:1]: Ah, see my other comment about optional dependencies - IMO extras are a good place for this to avoid adding it by default :)

File: gitlab/clients/httpx_client.py
Comment (nejch) [reply gitlab/clients/httpx_client.py]: Ah ok :) I'd personally go for `httpx.py` and `requests.py` as we already have the namespace.

File: gitlab/client.py
Comment (lmilbaum) [reply gitlab/client.py:23]: IMO, it is in the scope of this PR.  This is how a maintainer can get confidence that the contributed change is not breaking something that worked before. Will work on it.

File: requirements.txt
Comment (lmilbaum) [reply requirements.txt:1]: Yeah. Pushed a fix.

File: gitlab/clients/httpx_client.py
Comment (lmilbaum) [reply gitlab/clients/httpx_client.py]: I like the file name to be aligned with the class name

File: gitlab/client.py
Comment (lmilbaum) [reply gitlab/client.py:21]: @nejch I added a new workflow which installs the package and test whether `httpx` package can be imported or not. Not sure if this covers what you mentioned in your first comment. WDYT?

**Comment by lmilbaum** [comment]
@nejch The PR is waiting for review quite some time. I can understand that it is not a trivial PR. How about me setting a call for a merge party?

**Comment by nejch** [comment]
> @nejch The PR is waiting for review quite some time. I can understand that it is not a trivial PR. How about me setting a call for a merge party?

Sorry for the delay @lmilbaum, I was away a few days and there was a backlog of smaller PRs/issues :sweat_smile: 

I have a few pending comments and I wanted to have a closer look before finishing review as it's a larger change. We can also set up a call for later this week, sure, although I'll try to do another pass in the meantime. I'll send my pending comments for now.

File: gitlab/client.py
Comment (nejch) [inline gitlab/client.py:110]: Is it necessary that we offload all of this to the backend clients? It mostly just sets up auth headers, so ideally we could keep most of the logic in the existing client.

I have a suspicion some of this is needed because of `HTTPBasicAuth` in `_set_auth_info()`. In fact, this is sort of dead code as GitLab has long disabled basic auth via username/password. So instead we could perhaps remove that part of the code. WDYT?

File: setup.py
Comment (lmilbaum) [reply setup.py:56]: Makes sense.

File: tests/install/test_install.py
Comment (lmilbaum) [reply tests/install/test_install.py:7]: Right. I haven't used it for a while. Thanks.

File: gitlab/client.py
Comment (lmilbaum) [reply gitlab/client.py]: I would suggest to break it into two steps. Right now to do the minimal change of moving the same code from the general client to the requests client and do some cleanup in a follow up PR.

File: gitlab/client.py
Comment (JohnVillalovos) [inline gitlab/client.py:388]: Why is this method moved? It makes  it harder to review the changes, IMHO.

File: gitlab/client.py
Comment (JohnVillalovos) [inline gitlab/client.py:414]: Why is this method moved? It makes  it harder to review the changes, IMHO.

File: gitlab/clients/_httpxclient.py
Comment (JohnVillalovos) [inline gitlab/clients/_httpxclient.py:29]: Should there be an `__exit__` method?

File: gitlab/client.py
Comment (JohnVillalovos) [inline gitlab/client.py:29]: Maybe this could use the one in the requestsclient.py file or vice-versa.

Or move it to the constants file. Basically only have one copy of it, if possible.

File: gitlab/mixins.py
Comment (JohnVillalovos) [inline gitlab/mixins.py:59]: Why is the `Any` needed? And for all the other ones.

File: gitlab/clients/_httpxclient.py
Comment (JohnVillalovos) [inline gitlab/clients/_httpxclient.py:2]: Not sure what others think. But I would prefer the name to be either `_httpx_client` or `_httpx`. And the same idea on the current `_requestsclient`. But others may hate that idea!

**Review by JohnVillalovos — CHANGES_REQUESTED** [review]
@lmilbaum Thanks for all the work on this!!

I just did a quick review. If you disagree with my comments don't be afraid to say so 🙂 I am often wrong!

One concern I have is all the adding of `Any` to a lot of the type-hints.

**Comment by lmilbaum** [comment]
> @lmilbaum Thanks for all the work on this!!
> 
> I just did a quick review. If you disagree with my comments don't be afraid to say so 🙂 I am often wrong!
> 
> One concern I have is all the adding of `Any` to a lot of the type-hints.

Yeah. I can understand why. I didn't find any other solution. The understanding that this is an interim step, helps me go for this option.

File: .github/workflows/functional.yml
Comment (lmilbaum) [reply .github/workflows/functional.yml]: Right :-). Changed

File: gitlab/client.py
Comment (lmilbaum) [reply gitlab/client.py:73]: Yeah. Yet another agreement :-)

File: .pre-commit-config.yaml
Comment (lmilbaum) [reply .pre-commit-config.yaml]: By mistake. Removed.

File: gitlab/client.py
Comment (lmilbaum) [reply gitlab/client.py:416]: I agree. It is probably because I moved few function from this file to another.

File: gitlab/mixins.py
Comment (lmilbaum) [reply gitlab/mixins.py:59]: Because the httpx client function will return different type which I didn't want to set at this stage. Right now it is a placeholder for the next implementation phase.

File: gitlab/clients/_httpxclient.py
Comment (lmilbaum) [reply gitlab/clients/_httpxclient.py]: I don't have any strong feeling to either option.

**Comment by nejch** [comment]
>> @lmilbaum Thanks for all the work on this!!
>> I just did a quick review. If you disagree with my comments don't be afraid to say so slightly_smiling_face I am often wrong!
>> One concern I have is all the adding of Any to a lot of the type-hints.
>
> Yeah. I can understand why. I didn't find any other solution. The understanding that this is an interim step, helps me go for this option.

Thanks both for driving this, sorry again this is taking a while for me to get back into, I thought this would be easier and only the most low-level stuff factored out, similar to how it's done in https://python-redmine.com/advanced/request_engines.html /  https://github.com/maxtepkeev/python-redmine/tree/master/redminelib/engines. So just wanted to make sure we take a good look before we merge if there's a simple way :)

**Comment by lmilbaum** [comment]
> > > @lmilbaum Thanks for all the work on this!!
> > > I just did a quick review. If you disagree with my comments don't be afraid to say so slightly_smiling_face I am often wrong!
> > > One concern I have is all the adding of Any to a lot of the type-hints.
> > 
> > 
> > Yeah. I can understand why. I didn't find any other solution. The understanding that this is an interim step, helps me go for this option.
> 
> Thanks both for driving this, sorry again this is taking a while for me to get back into, I thought this would be easier and only the most low-level stuff factored out, similar to how it's done in https://python-redmine.com/advanced/request_engines.html / https://github.com/maxtepkeev/python-redmine/tree/master/redminelib/engines. So just wanted to make sure we take a good look before we merge if there's a simple way :)

I liked the approach taken in `python-redmine`.  The question I have with this approach is wether it goes along with our approach of eventually dropping the `request` http client/engine and stay with only `httpx`.

**Comment by lmilbaum** [comment]
Would breaking this PR into smaller pieces would make sense?

**Comment by nejch** [comment]
> Would breaking this PR into smaller pieces would make sense?

I think this might help yes @lmilbaum!
It would probably be a smaller PR if we don't take out all the methods as @JohnVillalovos suggested:

https://github.com/python-gitlab/python-gitlab/pull/2336/files#r1010073025
https://github.com/python-gitlab/python-gitlab/pull/2336/files#r1010073368

Also, it would potentially be a smaller PR if we ensure we keep public `http_*` methods in the main client so that we don't need to change the tests, because this is a public API on the Gitlab instance. We may just have private methods in backend clients that these wrap, but we somehow need to support this for now IMO.

https://python-gitlab.readthedocs.io/en/stable/api-levels.html

**Comment by JohnVillalovos** [comment]
So biggest issue with this for me is the type-hints. Everything having `Any` added. I worked a LONG time adding type-hints to the code and this kind of throws it all away :(

So I think we need to figure out a solution that doesn't use `Any`.

**Comment by lmilbaum** [comment]
Thank you for the feedback @nejch and @JohnVillalovos. I've started breaking this up into smaller PRs. Take a look at #2382

**Pull request closed without merging**
